---
title: "Comparing different conditions with scRNA-seq"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---

# Description
It's becoming more common that people are interested in performing experiments containing multiple experimental conditions. Handling this can be tricky. We'll talk about data integration and differential expression between experimental conditions here.

# Load package dependencies

```{r}
library(Seurat)
library(SeuratData)
library(tidyverse)
```

# Load the Seurat object

We'll load some data from the SeuratData package that contains IFNB-Stimulated and Control PBMCs

```{r}
# data(ifnb)
```

The loaded data isn't processed at all, so we'll fly through that

```{r}
seurat <- ifnb
seurat <- UpdateSeuratObject(seurat)
seurat <- NormalizeData(seurat)
seurat <- FindVariableFeatures(seurat)
seurat <- ScaleData(seurat)
seurat <- RunPCA(seurat, verbose=F)
seurat <- RunUMAP(seurat, dims=1:30)
seurat <- FindNeighbors(seurat, dims=1:30)
seurat <- FindClusters(seurat, resolution=0.3)
```

# Quick exploration
```{r}
DimPlot(seurat, label=T)
```

```{r}
DimPlot(seurat, group.by="seurat_annotations")
```

Whoa, there are two clusters of every cell type. Cool, right?!

Not really.

```{r}
DimPlot(seurat, group.by="stim")
```

So how do we test the effects of stimulation on cell types? Some clusters contain both conditions (cluster 4 = NK & CD8 T), while others span multiple clusters (Cluster 0 & 3 = CD4 Naive T).

# Integration in Seurat v5

This new section updates the integration analysis to use a new method introduced with Seurat v5. The original integration

The goal of integration is to find an embedding that aligns samples by a specific experimental or technical variable, negating its effects on the embedding. There are multiple methods for doing this, but here we'll stay in the Seurat ecosystem and use their approach.

In Seurat versions prior to v5, integration required two separate Seurat objects to be integrated; in Seurat v5 we can split the RNA assay into two `layers`, repeat the standard analysis pipeline on the split layers, then create the new embedding from the split assay.

```{r}
    seurat[["RNA"]] <- split(seurat[["RNA"]],f=seurat$stim)
```

We need to repeat the steps up to PCA on the split object prior to integrating

```{r}
seurat <- NormalizeData(seurat)
seurat <- FindVariableFeatures(seurat)
seurat <- ScaleData(seurat)
seurat <- RunPCA(seurat, verbose=F)
```

Having split the assay and created a new PCA reduction, a single command runs the integration. We're going to use the `CCA` method (canonical ... analysis) to integrate the data, but the streamlined integration approach in v5 offers several methods for integrating the samples:

# Anchor-based CCA integration (method=CCAIntegration)
# Anchor-based RPCA integration (method=RPCAIntegration)
# Harmony (method=HarmonyIntegration)
# FastMNN (method= FastMNNIntegration)
# scVI (method=scVIIntegration)

The [Seurat v5 integration vignette](https://satijalab.org/seurat/articles/seurat5_integration) on the Seurat website provides more details.

```{r}
seurat <- IntegrateLayers(obj=seurat,
                          method=CCAIntegration,
                          orig.reduction="pca",
                          new.reduction="integrated.cca")
```

We have added the `integrated.cca` dimensional reduction to the Seurat object - now we can repeat UMAP generation and clustering on this assay.

```{r}
seurat <- RunUMAP(seurat,reduction="integrated.cca",dims=1:30,reduction.name="umap.cca")
```

```{r}
DimPlot(seurat,reduction="umap.cca",label=T)
```

The position of cells in the UMAP projection has changed, and the clusters we identified earlier overlap with each other. Continue the process of finding clusters in the `integrated.cca` dimensional reduction and show these new clusters on the UMAP plot.

```{r}
seurat<-FindNeighbors(seurat,reduction="integrated.cca",dims=1:30)
seurat<-FindClusters(seurat,resolution=0.3)
```

```{r}
DimPlot(seurat, reduction="umap.cca", label=T)
DimPlot(seurat, reduction="umap.cca", group.by="stim")
DimPlot(seurat, reduction="umap.cca", group.by="seurat_annotations", label=T)
```

Now our clusters correspond to individual cell types and are not separated by experimental condition. This is great because now we can take each cluster and do differential expression between conditions and this will tell us how each cell type is affected by stimulation. We'll get to that in a second.

Integration is great to define unified clusters that aren't driven by experimental/technical variables. If this variable is an uninteresting technical feature (eg. batch effect), we can just continue with using this UMAP. But often in papers, we want unified clusters, but want our UMAPs to still show the effect of our experimental condition. 

Because we saved our new 'CCA' UMAP with a different name, both the original and new UMAP projections are stored in our seurat object (named `umap` and `umap.cca`). We can show our new (CCA integrated) clusters on the original UMAP projection without having to re-run `ScaleData`, `RunPCA`, and `RunUMAP`.

```{r}
DimPlot(seurat, reduction="umap", group.by="stim")
DimPlot(seurat,reduction="umap")
```

We've maintained the unified clusters, but the treatment effect is also retained in the UMAP.

## Differential expression between conditions

Integration in earlier Seurat releases required creation of a new 'integrated' assay that contained modified expression values; when this was done we had to be careful to switch back to the `RNA` assay prior to calculating expression differences between the conditions.

With Seurat v5 we haven't added a new assay to the object, but we did split the RNA assay into two parts. To calculate differentially-expressed genes between the treatments for each cell type we need to re-join the RNA data into a single layer prior to running the differential expression test.
```{r}
# DefaultAssay(seurat) <- "RNA"
seurat #print out the seurat object to show the split layers

seurat<-JoinLayers(seurat)

seurat # Re-print the object to show that layers have been joined.
```

With n=1 for the conditions, we are a little limited to what types of tests we can do, but the most common approach is likely to use a Wilcoxon test between conditions, treating each cell as an individual measurement. **If you have replicates for conditions,** best practice is probably to perform differential expression using standard bulk RNA-seq methods on pseudobulk data (ie. summing up all gene counts across cells of a replicate in a specific cluster). Check out the R package [muscat](https://www.nature.com/articles/s41467-020-19894-4) for this. It performs pseudobulking and implements a variety of standard methods (DESeq, edgeR, etc).

Let's just proceed with our n=1. Seurat's FindMarkers() function can perform the Wilcoxon test for us. I'll just perform differential expression for a single cell type / cluster (CD14 monocytes; cluster 0) here, but I'll include a function below that will iterate through this for all clusters

```{r}
mono_dge <- FindMarkers(seurat, ident.1="STIM", ident.2="CTRL",
                        group.by="stim", #the metadata column that has experimental condition
                        subset.ident = 0, #what cluster to subset and run this on
                        logfc.threshold = 0.25, only.pos=F)
```

Here's a chunk of code if you want to run this differential expression across all clusters. We'll skip it for now
```{r}
#runDGE <- function(cluster){
#  print(paste0("Running DGE for cluster: ", cluster))
#  results <- FindMarkers(seurat, ident.1="STIM", ident.2="CTRL",
#                         group.by="stim",
#                         subset.ident = cluster,
#                         logfc.threshold = 0.25,
#                         only.pos=F)
#  results$Cluster <- cluster
#  return(results)
#}

#cluster_list <- unique(seurat$seurat_clusters)
#dge_results <- lapply(cluster_list, runDGE)
#dge_results <- do.call("rbind", dge_results)
```


```{r}
mono_dge %>%
  arrange(desc(avg_log2FC)) %>%
  head()
```

```{r}
FeaturePlot(seurat, features="CCL8", cols=c('lightgrey', 'red'))
```

```{r}
VlnPlot(seurat, features="CCL8",
        split.by="stim", idents=0)
```

